cloud/ec2_fleet.go

```diff
@@ -0,0 +1,538 @@
+		if err != nil {
+			return nil, errors.Wrap(err, "not spawning host, problem creating key")
+		}
+		ec2Settings.KeyName = k
+	}
+
+	blockDevices, err := makeBlockDeviceMappingsTemplate(ec2Settings.MountPoints)
+	if err != nil {
+		return nil, errors.Wrap(err, "error making block device mappings")
+	}
+
+	r, err := getRegion(h)
+	if err != nil {
+		return nil, errors.Wrap(err, "problem getting region from host")
+	}
+	if err = m.client.Create(m.credentials, r); err != nil {
+		return nil, errors.Wrap(err, "error creating client")
+	}
+	defer m.client.Close()
+
+	resources, err := m.spawnFleetSpotHost(ctx, h, ec2Settings, blockDevices)
+	if err != nil {
+		msg := "error spawning spot host with Fleet"
+		grip.Error(message.WrapError(err, message.Fields{
+			"message":       msg,
+			"host":          h.Id,
+			"host_provider": h.Distro.Provider,
+			"distro":        h.Distro.Id,
+		}))
+		return nil, errors.Wrap(err, msg)
+	}
+	grip.Debug(message.Fields{
+		"message":       "spawned spot host with Fleet",
+		"host":          h.Id,
+		"host_provider": h.Distro.Provider,
+		"distro":        h.Distro.Id,
+	})
+
+	if err = setTags(ctx, resources, h, m.client, m.credentials); err != nil {
+		grip.Error(message.WrapError(err, message.Fields{
+			"message":       "error attaching tags",
+			"host":          h.Id,
+			"host_provider": h.Distro.Provider,
+			"distro":        h.Distro.Id,
+		}))
+		return nil, errors.Wrapf(err, "failed to attach tags for %s", h.Id)
+	}
+	grip.Debug(message.Fields{
+		"message":       "attached tags for host",
+		"host":          h.Id,
+		"host_provider": h.Distro.Provider,
+		"distro":        h.Distro.Id,
+	})
+
+	event.LogHostStarted(h.Id)
+	return h, nil
+}
+
+func (m *ec2FleetManager) GetInstanceStatuses(ctx context.Context, hosts []host.Host) ([]CloudStatus, error) {
+	instanceIDs := make([]*string, 0, len(hosts))
+	for _, h := range hosts {
+		instanceIDs = append(instanceIDs, aws.String(h.Id))
+	}
+
+	if err := m.client.Create(m.credentials, defaultRegion); err != nil {
+		return nil, errors.Wrap(err, "error creating client")
+	}
+	defer m.client.Close()
+
+	describeInstancesOutput, err := m.client.DescribeInstances(ctx, &ec2.DescribeInstancesInput{
+		InstanceIds: instanceIDs,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "error describing instances")
+	}
+
+	if err = validateEc2DescribeInstancesOutput(describeInstancesOutput); err != nil {
+		return nil, errors.Wrap(err, "invalid describe instances response")
+	}
+
+	if len(instanceIDs) != len(describeInstancesOutput.Reservations) {
+		return nil, errors.Errorf("AWS returned %d statuses for %d hosts", len(describeInstancesOutput.Reservations), len(instanceIDs))
+	}
+
+	statusMap := map[string]CloudStatus{}
+	for i := range describeInstancesOutput.Reservations {
+		statusMap[*describeInstancesOutput.Reservations[i].Instances[0].InstanceId] = ec2StatusToEvergreenStatus(*describeInstancesOutput.Reservations[i].Instances[0].State.Name)
```

Comment:
this feels risky, but maybe there's nothing to be done.

Reply:
we do validate the response from AWS to make sure this won't panic.
What specifically is risky?